### PR TITLE
Strengthen well known check-cfg names and values test

### PR DESCRIPTION
**Pull request by Urgau**
https://github.com/rust-lang/rust/pull/118494 is changing the implementation of how we expect well known check-cfg names and values, but we currently don't have a test that checks every well known only some of them.

This PR therefore strengthen our well known names/values test to include all of the configs to at least avoid unintended regressions and validate new entry.

*this PR also contains some drive-by consolidation of unexpected `target_os`, `target_arch` into a single file*

r? @nnethercote (maybe? feel free to re-assign)

### CODE REVIEW

File: compiler/rustc_session/src/config.rs [hunk 1]
```diff
@@ -1422,6 +1422,9 @@ impl CheckCfg {
         };
 
         // NOTE: This should be kept in sync with `default_configuration`
+        //
+        // When adding a new config here you should also update
+        // `tests/ui/check-cfg/well-known-values.rs`.
```
Comment (nnethercote) [inline compiler/rustc_session/src/config.rs:1427]: 👍

File: tests/ui/check-cfg/well-known-values.rs [hunk 2]
```diff
@@ -1,41 +1,90 @@
-// This test check that we lint on non well known values and that we don't lint on well known
-// values
+// This test check that we recogniez all the well known config names
```
Comment (nnethercote) [inline tests/ui/check-cfg/well-known-values.rs:1]:
```suggestion
// This test check that we recognize all the well known config names
```

File: tests/ui/check-cfg/well-known-values.rs [hunk 3]
```diff
@@ -1,41 +1,90 @@
...
+#![feature(cfg_target_has_atomic)]
+#![feature(cfg_target_has_atomic_equal_alignment)]
+#![feature(cfg_target_abi)]
+#![feature(cfg_relocation_model)]
+#![feature(cfg_sanitize)]
+#![feature(cfg_target_thread_local)]
+
+#[cfg(any(
+    doc = "_UNEXPECTED_VALUE",
+    //~^ WARN unexpected `cfg` condition value
```
Comment (nnethercote) [inline tests/ui/check-cfg/well-known-values.rs:20]: I would put this list in alphabetical order, and maybe even add `tidy-alphabetical-{start,end}` markers so it is guaranteed to stay in order.

File: tests/ui/check-cfg/well-known-values.rs [hunk 4]
```diff
@@ -1,41 +1,90 @@
...
 
 #[cfg(target_os = "linux")]
 fn target_os_linux() {}
 
-#[cfg(target_has_atomic = "0")]
-//~^ WARNING unexpected `cfg` condition value
-fn target_has_atomic_invalid() {}
-
 #[cfg(target_has_atomic = "8")]
 fn target_has_atomic() {}
```
Comment (nnethercote) [inline tests/ui/check-cfg/well-known-values.rs:80]: It's not clear to me why some of these remain, when they are now mentioned in the big list above? But some were removed.

Comment (nnethercote) [reply tests/ui/check-cfg/well-known-values.rs:80]: The linux_misspell one makes sense to remain, but I don't understand the rest.

**Comment by nnethercote** [comment]
> this PR also contains some drive-by consolidation of unexpected target_os, target_arch into a single file

I generally put these kinds of changes into separate commits. It makes the reviewer's job easier when unrelated changes are separated out.

File: tests/ui/check-cfg/well-known-values.rs (hunk 4 above)
Comment (Urgau) [reply tests/ui/check-cfg/well-known-values.rs:93]: We still need to test that the combination of expected well known names and values do not produce a warning, but we do not need to test all of them, so I chose some that seemed relevant (with/without a value, both at the same time). It's very arbitrary and I'm happy to change it if desired.

Comment (nnethercote) [reply tests/ui/check-cfg/well-known-values.rs:93]: A comment explaining exactly that would be helpful :)

Comment (Urgau) [reply tests/ui/check-cfg/well-known-values.rs:99]: Done. I've improved the comment at the start of the section.

File: tests/ui/check-cfg/well-known-values.rs (hunk 3 above)
Comment (Urgau) [reply tests/ui/check-cfg/well-known-values.rs]: Great, and to make sure it works as expected I've tested it locally with success.

File: tests/ui/check-cfg/well-known-values.rs [hunk 5]
```diff
@@ -1,41 +1,106 @@
...
+#![feature(cfg_target_thread_local)]
+
+// This part makes sure that none of the well known names are
+// unexpected.
+//
+// BUT to make sure that no expected values changes without
+// being noticed we pass them a obviously wrong value so the
+// diagnostic prints the list of expected values.
+//
+// This is in alphabetical order.
```
Comment (nnethercote) [inline tests/ui/check-cfg/well-known-values.rs:25]: Don't need this line any more now that the tidy markers are present.

With that, I think it's ready to go.

@bors delegate=Urgau

**Comment by GuillaumeGomez** [comment]
Weird. Oh well.

@bors r=nnethercote rollup

**Comment by GuillaumeGomez** [comment]
Failed in https://github.com/rust-lang/rust/pull/118744.

**Comment by Urgau** [comment]
I've rebased and re-blessed the diagnostic output (new win7 target).

@rustbot ready

**Pull request merged**